demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -85,129 +145,344 @@ typedef struct
+/** @brief Static buffer used to hold MQTT messages being sent and received. */
+static MQTTFixedBuffer_t xBuffer =
+{
+    ucSharedBuffer,
+    NETWORK_BUFFER_SIZE
+};
+
+static void _sendMessageToGGC( MQTTContext_t * pxMQTTContext );
+static int _discoverGreengrassCore();
+static void prvEventCallback( MQTTContext_t * pxMQTTContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              MQTTDeserializedInfo_t * pxDeserializedInfo );
+static uint32_t prvGetTimeMs( void );
+static BaseType_t prvBackoffForRetry( BackoffAlgorithmContext_t * pxRetryParams );
+static BaseType_t prvCreateMQTTConnectionWithBroker( MQTTContext_t * pxMQTTContext,
+                                                     NetworkContext_t * pxNetworkContext );
 
 /*-----------------------------------------------------------*/
+static void prvEventCallback( MQTTContext_t * pxMQTTContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              MQTTDeserializedInfo_t * pxDeserializedInfo )
+{
+    return;
+}
 
-static void _sendMessageToGGC( IotMqttConnection_t mqttConnection )
+static uint32_t prvGetTimeMs( void )
 {
-    const char * pcTopic = ggdDEMO_MQTT_MSG_TOPIC;
-    uint32_t ulMessageCounter;
-    char cBuffer[ ggdDEMO_MAX_MQTT_MSG_SIZE ];
-    IotMqttError_t xMqttStatus = IOT_MQTT_STATUS_PENDING;
-    IotMqttPublishInfo_t xPublishInfo = IOT_MQTT_PUBLISH_INFO_INITIALIZER;
+    TickType_t xTickCount = 0;
+    uint32_t ulTimeMs = 0UL;
 
+    /* Get the current tick count. */
+    xTickCount = xTaskGetTickCount();
 
-    /* Publish to the topic to which this task is subscribed in order
-     * to receive back the data that was published. */
-    xPublishInfo.qos = IOT_MQTT_QOS_0;
-    xPublishInfo.pTopicName = pcTopic;
-    xPublishInfo.topicNameLength = ( uint16_t ) ( strlen( pcTopic ) );
+    /* Convert the ticks to milliseconds. */
+    ulTimeMs = ( uint32_t ) xTickCount * MILLISECONDS_PER_TICK;
 
-    for( ulMessageCounter = 0; ulMessageCounter < ( uint32_t ) ggdDEMO_MAX_MQTT_MESSAGES; ulMessageCounter++ )
-    {
-        /* Set the members of the publish info. */
-        xPublishInfo.payloadLength = ( uint32_t ) sprintf( cBuffer, ggdDEMO_MQTT_MSG_DISCOVERY, ( long unsigned int ) ulMessageCounter ); /*lint !e586 sprintf can be used for specific demo. */
-        xPublishInfo.pPayload = ( const void * ) cBuffer;
+    /* Reduce ulGlobalEntryTimeMs from obtained time so as to always return the
+     * elapsed time in the application. */
+    ulTimeMs = ( uint32_t ) ( ulTimeMs - ulGlobalEntryTimeMs );
 
-        /* Call the MQTT blocking PUBLISH function. */
-        xMqttStatus = IotMqtt_TimedPublish( mqttConnection,
-                                            &xPublishInfo,
-                                            0,
-                                            _maxCommandTimeMs );
+    return ulTimeMs;
+}
+static BaseType_t prvBackoffForRetry( BackoffAlgorithmContext_t * pxRetryParams )
```

Comment:
It looks like a good chunk of the functions here can just reference the code in https://github.com/aws/amazon-freertos/tree/main/demos/common/mqtt_demo_helpers

It might require slight modifications to the CMake but referencing common functions from one place will make it easier to make bug fixes in the future

Reply:
As discussed we can have it in the separate PR.

Reply:
Publish and disconnect at least can use the common code here

Reply:
Logic in Disconnect helper function is based on some static variables which are set in mqttConnection helper function. So it won't work it in this demo. The publish helper function publishes QOS1 messages whereas in ggd demo we are showing QOS0, so it will unnecessarily complicate the demo. May be I can include QOS1 changes in the PR where I will be using the demo helper functions.